oit-dual-depth-peeling.html

```diff
@@ -0,0 +1,722 @@
+                gl.clearColor(0, 0, 0, 0);
+                gl.clear(gl.COLOR_BUFFER_BIT);
+
+                gl.bindFramebuffer(gl.FRAMEBUFFER, allBuffers[0]);
+                gl.drawBuffers([gl.COLOR_ATTACHMENT0]);
+                gl.clearColor(DEPTH_CLEAR_VALUE, DEPTH_CLEAR_VALUE, 0, 0);
+                gl.clear(gl.COLOR_BUFFER_BIT);
+
+                gl.bindFramebuffer(gl.DRAW_FRAMEBUFFER, allBuffers[1]);
+                // gl.drawBuffers([gl.COLOR_ATTACHMENT0]);
+                gl.clearColor(-MIN_DEPTH, MAX_DEPTH, 0, 0);
+                gl.clear(gl.COLOR_BUFFER_BIT);
+
+                gl.bindFramebuffer(gl.DRAW_FRAMEBUFFER, colorBuffers[0]);
+                gl.drawBuffers([gl.COLOR_ATTACHMENT0, gl.COLOR_ATTACHMENT1]);
+                gl.clearColor(0, 0, 0, 0);
+                gl.clear(gl.COLOR_BUFFER_BIT);
+
+                gl.bindFramebuffer(gl.DRAW_FRAMEBUFFER, colorBuffers[1]);
+                gl.drawBuffers([gl.COLOR_ATTACHMENT0, gl.COLOR_ATTACHMENT1]);
+                gl.clearColor(0, 0, 0, 0);
+                gl.clear(gl.COLOR_BUFFER_BIT);
+
+                 // draw depth for first pass
+                gl.bindFramebuffer(gl.DRAW_FRAMEBUFFER, allBuffers[0]);
+                gl.drawBuffers([gl.COLOR_ATTACHMENT0]);
+                gl.blendEquation(gl.MAX);
+
+                gl.useProgram(dualDepthPeelingProgram);
+                gl.uniform1i(dualDepthPeelingLocations.depth, 3);
+                gl.uniform1i(dualDepthPeelingLocations.frontColor, 4);
+                gl.bindVertexArray(sphereArray);
+
+                for (var i = 0, len = spheres.length; i < len; ++i) {
+                    spheres[i].rotate[1] += 0.002;
+
+                    utils.xformMatrix(spheres[i].modelMatrix, spheres[i].translate, null, spheres[i].scale);
+                    mat4.fromYRotation(rotationMatrix, spheres[i].rotate[1]);
+                    mat4.multiply(spheres[i].modelMatrix, rotationMatrix, spheres[i].modelMatrix);
+                
+                    modelMatrixData.set(spheres[i].modelMatrix, i * 16);
+                }
+
+                gl.bindBuffer(gl.ARRAY_BUFFER, matrixBuffer);
+                gl.bufferSubData(gl.ARRAY_BUFFER, 0, modelMatrixData);
+                
+                gl.drawElementsInstanced(gl.TRIANGLES, sphere.indices.length, gl.UNSIGNED_SHORT, 0, spheres.length);
+
+
+                ////////////////////////////////////
+                // 2. Dual Depth Peeling Ping-Pong
+                ////////////////////////////////////
+                let readId, writeId;
+                let offsetRead, offsetBack;
+
+                for (let pass = 0; pass < NUM_PASS; pass++) {
+                    readId = pass % 2;
+                    writeId = 1 - readId;  // ping-pong: 0 or 1
+                    
+
+                    gl.bindFramebuffer(gl.DRAW_FRAMEBUFFER, allBuffers[writeId]);
+                    gl.drawBuffers([gl.COLOR_ATTACHMENT0]);

```

Comment:
This might just be my understanding of drawBuffers, but I don't think you need to call this before each clear? Can you not just set the drawbuffers when you initialize your framebuffers, and then not repeat the calls here? Also, clear the color buffer first, so you don't have to bind the "allBuffer" twice.

Reply:
clearing colorBuffers first wouldn't work. It seems that even if I already set gl.drawBuffers([gl.COLOR_ATTACHMENT0]); for allBuffers, It's still affecting the other two color attachment...
I think gl.drawBuffers still needs calling if it changes. I've removed those that are same as current draw buffer states